core/src/main/java/com/linecorp/armeria/common/stream/HttpDeframer.java

```diff
@@ -0,0 +1,274 @@
+    private static final AtomicReferenceFieldUpdater<HttpDeframer, Subscription> upstreamUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(HttpDeframer.class, Subscription.class, "upstream");
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<HttpDeframer> initializedUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(HttpDeframer.class, "initialized");
+
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<HttpDeframer> subscribedUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(HttpDeframer.class, "subscribed");
+
+    private final ByteBufDeframerInput input;
+    private final ArrayDeque<T> outputQueue = new ArrayDeque<>();
+
+    private boolean sawLeadingHeaders;
+
+    @Nullable
+    private volatile EventExecutor eventLoop;
+    @Nullable
+    private volatile Subscription upstream;
+    private volatile boolean cancelled;
+    private volatile int initialized;
+    private volatile int subscribed;
+
+    /**
+     * Returns a new {@link HttpDeframer} with the specified {@link EventExecutor} and
+     * {@link ByteBufAllocator}.
+     */
+    protected HttpDeframer(ByteBufAllocator alloc) {
+        requireNonNull(alloc, "alloc");
+        input = new ByteBufDeframerInput(alloc);
+    }
+
+    /**
+     * Converts the specified {@link HttpData} to {@link ByteBuf}.
+     */
+    protected ByteBuf convertToByteBuf(HttpData data) {
+        return data.byteBuf();
+    }
+
+    /**
+     * Decodes a stream of {@link HttpData}s to N objects.
+     * This method will be called whenever an {@link HttpData} is signaled from {@link Publisher}.
+     */
+    protected abstract void process(HttpDeframerInput in, HttpDeframerOutput<T> out);
+
+    private void process(HttpObject data) {
+        if (data instanceof HttpHeaders) {
+            final HttpHeaders headers = (HttpHeaders) data;
+
+            if (headers instanceof ResponseHeaders) {
+                final ResponseHeaders responseHeaders = (ResponseHeaders) headers;
+                if (responseHeaders.status().isInformational()) {
+                    processInformationalHeaders(responseHeaders, outputQueue::addLast);
+                    return;
+                }
+            }
+
+            if (!sawLeadingHeaders) {
+                sawLeadingHeaders = true;
+                processHeaders((HttpHeaders) data, outputQueue::addLast);
+            } else {
+                processTrailers((HttpHeaders) data, outputQueue::addLast);
+            }
+            return;
+        }
+
+        if (data instanceof HttpData) {
+            input.add(convertToByteBuf((HttpData) data));
+            process(input, outputQueue::addLast);
+            input.discardReadBytes();
+        }
+    }
+
+    /**
+     * Decodes an informational {@link ResponseHeaders} to N objects.
+     */
+    protected void processInformationalHeaders(ResponseHeaders in, HttpDeframerOutput<T> out) {}
```

Comment:
Should we prefer composition over inheritance? e.g. add a dedicated interface that provides these protected methods?